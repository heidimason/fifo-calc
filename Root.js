import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { combineReducers } from 'redux'

import PurchaseHistoryReducer from './src/reducers/PurchaseHistory'
import SaleHistoryReducer from './src/reducers/SaleHistory'

const rootReducer = combineReducers({
		purchaseHistory: PurchaseHistoryReducer,
		saleHistory: SaleHistoryReducer
	}),

	store = createStore(
		rootReducer
	)

export default props => {
	return (
		<Provider store={store}>
			{props.children}
		</Provider>
	)
}
