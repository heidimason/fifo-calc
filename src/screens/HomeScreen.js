import React, { Component } from 'react'
import { Platform, StatusBar, StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import styled from 'styled-components/native'
import { withNavigation } from 'react-navigation'
import { connect } from 'react-redux'
// import serializeForm from 'form-serialize'

import NumShares from '../components/NumShares'
import PriceOfShares from '../components/PriceOfShares'
import TextInputError from '../components/TextInputError'
import SubmitBtn from '../components/SubmitBtn'
import { isValidChar } from '../utils/helpers'
import { blue, grayDark, green, orange, red, white } from '../utils/styles/colors'
import { app } from '../utils/styles/app'
import { btns } from '../utils/styles/btns'
import { fonts } from '../utils/styles/fonts'
import { forms } from '../utils/styles/forms'
import { addPurchase } from '../actions/PurchaseHistory'
import { addSale } from '../actions/SaleHistory'

class HomeScreen extends Component {
	state = {
		isValidSaleNum: true,
		profit: 0,
		purchases: [],
		purchaseNum: '',
		purchasePrice: '',
		saleNum: '',
		salePrice: ''
	}

	changePurchaseShares = numShares => {
		this.setState({
			purchaseNum: numShares
		})
	}

	changePurchasePrice = pricePerShare => {
		this.setState({
			purchasePrice: pricePerShare
		})
	}

	submitPurchase = e => {
		// e.preventDefault()

		// const values = serializeForm(e.target, { hash: true })

		const { purchases, purchaseNum, purchasePrice } = this.state,
			 { purchaseHistory, updatePurchaseHistory } = this.props

		const purchase = {
			num: parseInt(purchaseNum),
			price: parseInt(purchasePrice)
		}

		// const purchaseCopy = Object.assign(values, {
		// 	index: parseInt(purchaseHistory.length + 1),
		// 	num: parseInt(purchaseNum),
		// 	price: parseInt(purchasePrice)
		// })

		const purchaseCopy = JSON.parse( JSON.stringify(purchase) )

		this.setState({
			purchases: [purchase, ...purchases]
		})

		updatePurchaseHistory(purchaseCopy)
	}

	toHistory = () => {
        const { profit } = this.state,
		  { navigation } = this.props

        navigation.navigate('History', {
        	profit
        })
    }

	changeSaleShares = numShares => {
		this.setState({
			saleNum: numShares
		})
	}

	changeSalePrice = pricePerShare => {
		this.setState({
			salePrice: pricePerShare
		})
	}

	calculateProfit = (purchasesParam, saleNumParam, salePriceParam, profitParam = 0 || this.state.profit) => {
		const {
			isValidSaleNum,
			profit,
			purchases,
			saleNum,
			salePrice
		} = this.state

		const { saleHistory } = this.props

		const salePriceInt = parseInt(salePriceParam),
		  purchasePriceInt = parseInt(purchasesParam[purchasesParam.length - 1].price)

		let saleNumInt = parseInt(saleNumParam),
		purchaseNumInt = parseInt(purchasesParam[purchasesParam.length - 1].num)

		// Base case
		if (saleNumInt === 0) {
			// For Sales History
			const saleCopy = {
				index: parseInt(saleHistory.length + 1),
				num: parseInt(saleNum),
				price: parseInt(salePrice)
			}

			this.setState({
				isValidSaleNum: true,
				profit: profitParam,
				purchases: purchasesParam,
				saleHistory: [saleCopy, ...saleHistory]
			})

			if (purchaseNumInt === 0) {
				purchasesParam.length--
			}

			return
		}

		// Recursively calculate profit
		if (saleNumInt <= purchaseNumInt) {
			profitParam += (saleNumInt * salePriceInt) - (saleNumInt * purchasePriceInt)

			purchasesParam[purchasesParam.length - 1].num -= saleNumInt

			saleNumParam -= saleNumInt

			return this.calculateProfit(purchasesParam, saleNumParam, salePriceParam, profitParam)
		} else {
			// Calculate batch by batch
			if (purchasesParam.length > 1) {
				profitParam += (purchaseNumInt * salePriceInt) - (purchaseNumInt * purchasePriceInt)

				purchasesParam[purchasesParam.length - 1].num -= purchaseNumInt

				saleNumParam -= purchaseNumInt

				if (purchasesParam[purchasesParam.length - 1].num === 0) {
					purchasesParam.length--
				}

				return this.calculateProfit(purchasesParam, saleNumParam, salePriceParam, profitParam)
			} else {
				// Display error text (i.e. "Number of sale shares must not exceed number of purchase shares!")
				this.setState({
					isValidSaleNum: false
				})
			}
		}
	}

	resetAll = () => {
		this.setState({
			isValidSaleNum: true,
			profit: 0,
			purchases: [],
			purchaseNum: '',
			purchasePrice: '',
			saleNum: '',
			salePrice: ''
		})
	}

	render () {
		StatusBar.setBarStyle('light-content', true)

		const {
			isValidSaleNum,
			profit,
			purchases,
			purchaseNum,
			purchasePrice,
			saleNum,
			salePrice
		} = this.state,

		{ purchaseHistory } = this.props

		return (
			<View style={app.container}>
				<TextContainer>
					<H1
						style={[fonts.h1, fonts.text]}>FIFO Profit Calculator
					</H1>
				</TextContainer>

				<View>
					<Text
						style={[fonts.h2, fonts.text, styles.h2]}>Purchase
					</Text>

					<NumShares
						onNumChange={purchaseNum => {
							this.changePurchaseShares(purchaseNum)
						}}
						shareNum={purchaseNum}
					/>

					{ purchaseNum !== '' && purchaseNum === '0' || !isValidChar(purchaseNum)
						? <TextInputError text="Please enter a positive integer only!" />
						: null
					}

					<PriceOfShares
						onPriceChange={purchasePrice => {
							this.changePurchasePrice(purchasePrice)
						}}
						sharePrice={purchasePrice}
					/>

					{ purchasePrice !== '' && purchasePrice === '0' || !isValidChar(purchasePrice)
						? <TextInputError text="Please enter a positive integer only!" />
						: null
					}
				</View>

				<SubmitBtn
					children="Submit"
					disabled={
						purchaseNum === '' || purchasePrice === '' ||
						!isValidChar(purchaseNum) || !isValidChar(purchasePrice)
					}
					onPress={this.submitPurchase}
					style={[
						Platform.OS === 'ios'
						? btns.btnIOS
						: btns.btnAndroid,
						purchaseNum === '' || purchasePrice === '' ||
						!isValidChar(purchaseNum) || !isValidChar(purchasePrice)
						? btns.btnInvalid
						: btns.btnValid,
						[btns.btn, styles.submitBtnPurchase]
					]}
				/>

				<HistoryBtn
					disabled={!purchaseHistory.length}
					onPress={this.toHistory}
					style={[
						!purchaseHistory.length
						? btns.btnInvalid
						: btns.btnValid,
						btns.btn
					]}>
					{ Platform.OS === 'ios' ?
						<HistoryText
							style={[btns.btnText, fonts.h3, {position: 'absolute'}]}>History
						</HistoryText>
						:
						<HistoryText
							style={[btns.btnText, fonts.h3]}>History
						</HistoryText>
					}
				</HistoryBtn>

				<View>
					<Text
						style={[fonts.h2, fonts.text, styles.h2]}>Sale
					</Text>

					<NumShares
						onNumChange={saleNum => {
							this.changeSaleShares(saleNum)
						}}
						shareNum={saleNum}
					/>

					{ saleNum !== '' && saleNum === '0' || !isValidChar(saleNum)
						? <TextInputError text="Please enter a positive integer only!" />
						: null
					}

					{ !isValidSaleNum
						? <TextInputError text="Number of sale shares must not exceed number of purchase shares!" />
						: null
					}

					<PriceOfShares
						onPriceChange={salePrice => {
							this.changeSalePrice(salePrice)
						}}
						sharePrice={salePrice}
					/>

					{ salePrice !== '' && salePrice === '0' || !isValidChar(salePrice)
						? <TextInputError text="Please enter a positive integer only!" />
						: null
					}
				</View>

				<SubmitBtn
					children="Calculate"
					disabled={
						!purchases.length ||
						saleNum === '' || salePrice === '' ||
						!isValidChar(saleNum) || !isValidChar(salePrice)
					}
					onPress={() => this.calculateProfit(purchases, saleNum, salePrice)}
					style={[
						Platform.OS === 'ios'
						? btns.btnIOS
						: btns.btnAndroid,
						!purchases.length ||
						saleNum === '' || salePrice === '' ||
						!isValidChar(saleNum) || !isValidChar(salePrice)
						? btns.btnInvalid
						: btns.btnValid,
						[btns.btn, styles.submitBtnCalculate]
					]}
				/>

				<TextContainer>
					<Text
						style={[fonts.h2, fonts.profit, fonts.text]}>Profit: {profit}
					</Text>
				</TextContainer>

				<SubmitBtn
					children="Reset All"
					disabled={
						!purchases.length &&
						purchaseNum === '' && saleNum === '' &&
						purchasePrice === '' && salePrice === ''}
					onPress={this.resetAll}
					style={[
						Platform.OS === 'ios'
						? btns.btnIOS
						: btns.btnAndroid,
						!purchases.length &&
						purchaseNum === '' && saleNum === '' &&
						purchasePrice === '' && salePrice === ''
						? btns.btnInvalid
						: btns.btnValid,
						[btns.btn, styles.submitBtnReset]
					]}
				/>
			</View>
		)
	}
}

const TextContainer = styled.View`
		align-items: center
	`,
	H1 = styled.Text`
		margin-top: 30
	`,
	HistoryBtn = styled.TouchableOpacity`
		align-self: flex-end
		top: 10
		transform: rotate(270deg)
	`,
	HistoryText = styled.Text`
		align-self: center
		background-color: ${green}
		letter-spacing: 3
		paddingHorizontal: 3
		text-transform: uppercase
		top: 3
	`

const styles = StyleSheet.create({
	h2: {
		marginBottom: 10
	},
	submitBtnPurchase: {
		backgroundColor: blue
	},
	submitBtnCalculate: {
		backgroundColor: orange
	},
	submitBtnReset: {
		backgroundColor: red,
		marginBottom: 30
	}
})

const mapStateToProps = state => {
    return {
        purchaseHistory: state.purchaseHistory,
        saleHistory: state.saleHistory
    }
}

const mapDispatchToProps = dispatch => {
    return {
        updatePurchaseHistory: purchase => dispatch( addPurchase(purchase) ),
        updateSaleHistory: sale => dispatch( addSale(sale) )
    }
}

export default withNavigation( connect(mapStateToProps, mapDispatchToProps)(HomeScreen) )
