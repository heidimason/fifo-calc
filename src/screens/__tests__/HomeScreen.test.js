import 'jsdom-global/register'
import React from 'react'
// import renderer from 'react-test-renderer'
import { mount, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import sinon from 'sinon'

import Root from '../../../Root'
import HomeScreen from '../HomeScreen'

configure({
	adapter: new Adapter
})

describe('<HomeScreen />', () => {
	let wrapper = mount(
		<Root>
			<HomeScreen />
		</Root>
	)

	// const tree = renderer.create(
	// 	<Root>
	// 		<HomeScreen />
	// 	</Root>
	// ).toJSON()

	// it('renders correctly', () => {
	// 	expect(tree).toMatchSnapshot()
	// })

	// it('has 1 child', () => {
	// 	expect(tree.children.length).toBe(8)
	// })
})

// describe('functions exist', () => {
// 	test('function changePurchaseShares exists', () => {
// 		expect(typeof changePurchaseShares).toEqual('function')
// 	})

// 	test('function changePurchasePrice exists', () => {
// 		expect(typeof changePurchasePrice).toEqual('function')
// 	})

// 	test('function submitPurchase exists', () => {
// 		expect(typeof submitPurchase).toEqual('function')
// 	})

// 	test('function changeSaleShares exists', () => {
// 		expect(typeof changeSaleShares).toEqual('function')
// 	})

// 	test('function changeSalePrice exists', () => {
// 		expect(typeof changeSalePrice).toEqual('function')
// 	})

// 	test('function calculateProfit exists', () => {
// 		expect(typeof calculateProfit).toEqual('function')
// 	})

// 	test('function resetAll exists', () => {
// 		expect(typeof resetAll).toEqual('function')
// 	})
// })

// describe('calculateProfit function', () => {
// 	let wrapper

// 	beforeEach(() => wrapper = shallow(<HomeScreen />))

// 	test('calculates profit correctly', () => {
// 		wrapper.setState({ profit: 0 })
// 		wrapper.setState({ purchases: ['num': 100, 'price': 25] })
// 		wrapper.setState({ purchaseNum: '100' })
// 		wrapper.setState({ purchasePrice: '25' })
// 	    wrapper.setState({ saleNum: '80' })
// 	    wrapper.setState({ salePrice: '35' })
// 	    wrapper.instance().calculateProfit()
// 	    expect(wrapper.state('profit')).toEqual('800')
// 	// 	expect( calculateProfit([{num: 100, price: 25}], 80, 35, 0) ).toBe(800)
// 	})
// })

// 1. Working
// Purchases
// 	100, 25

// Sale
// 80, 35
// Answer: 800

// 2. Working
// Purchases
// 	100, 25

// Sale
// 80, 10
// Answer: -1200

// 3. Working
// Purchases
// 	100, 25
// 100, 30

// Sale
// 80, 35
// Answer: 800


// 4. Not Working
// Purchases
// 	100, 25
// 100, 30

// Sale
// 120, 35

// $10 off first 100 shares = $1000
// $5 for remaining 20 shares = $100

// Answer: $1100

// 5. Not working
// Purchases
// 	100, 25
// 100, 30
// 100, 35

// Sale
// 250, 35
// Answer:

// 1st batch - $10 gain/share at 100 shares = $1000
// 2nd - $5/share for 100 shares = $500
// 3rd - $50 shares at 35 = $0

// Answer: $1500

// Note: Profit should only go below 0 if sale price < purchase price


