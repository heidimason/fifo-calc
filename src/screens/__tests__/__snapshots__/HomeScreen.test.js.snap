// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#222",
      "flex": 1,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "marginTop": 30,
          },
          Object {
            "fontSize": 53,
          },
          Object {
            "color": "#fff",
            "marginHorizontal": 40,
          },
        ]
      }
    >
      FIFO Profit Calculator
    </Text>
  </View>
  <View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 40,
          },
          Object {
            "color": "#fff",
            "marginHorizontal": 40,
          },
          Object {
            "marginBottom": 10,
          },
        ]
      }
    >
      Purchase
    </Text>
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      placeholder="Number of Shares"
      rejectResponderTermination={true}
      selectionColor="#222"
      style={
        Array [
          Object {
            "marginBottom": 15,
          },
          Object {
            "borderRadius": 7,
          },
          Object {
            "borderColor": "transparent",
            "borderWidth": 0,
          },
          Array [
            Object {
              "fontSize": 40,
            },
            Object {
              "backgroundColor": "#fff",
              "color": "#aaa",
              "height": 45,
              "marginHorizontal": 40,
              "textAlign": "center",
            },
          ],
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      placeholder="Price per Share"
      rejectResponderTermination={true}
      selectionColor="#222"
      style={
        Array [
          Object {
            "marginBottom": 15,
          },
          Object {
            "borderRadius": 7,
          },
          Object {
            "borderColor": "transparent",
            "borderWidth": 0,
          },
          Array [
            Object {
              "fontSize": 40,
            },
            Object {
              "backgroundColor": "#fff",
              "color": "#aaa",
              "height": 45,
              "marginHorizontal": 40,
              "textAlign": "center",
            },
          ],
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#0099ff",
        "borderRadius": 7,
        "height": 45,
        "justifyContent": "center",
        "marginHorizontal": 40,
        "opacity": 0.3,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontWeight": "bold",
            "textAlign": "center",
          },
          Object {
            "fontSize": 40,
          },
        ]
      }
    >
      Submit
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "flex-end",
        "height": 45,
        "opacity": 0.3,
        "top": 18,
        "transform": Array [
          Object {
            "rotate": "270deg",
          },
        ],
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "#228b22",
            "letterSpacing": 3,
            "paddingHorizontal": 3,
            "textTransform": "uppercase",
            "top": 3,
          },
          Object {
            "color": "#fff",
            "fontWeight": "bold",
            "textAlign": "center",
          },
          Object {
            "fontSize": 27,
          },
          Object {
            "position": "absolute",
          },
        ]
      }
    >
      History
    </Text>
  </View>
  <View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 40,
          },
          Object {
            "color": "#fff",
            "marginHorizontal": 40,
          },
          Object {
            "marginBottom": 10,
          },
        ]
      }
    >
      Sale
    </Text>
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      placeholder="Number of Shares"
      rejectResponderTermination={true}
      selectionColor="#222"
      style={
        Array [
          Object {
            "marginBottom": 15,
          },
          Object {
            "borderRadius": 7,
          },
          Object {
            "borderColor": "transparent",
            "borderWidth": 0,
          },
          Array [
            Object {
              "fontSize": 40,
            },
            Object {
              "backgroundColor": "#fff",
              "color": "#aaa",
              "height": 45,
              "marginHorizontal": 40,
              "textAlign": "center",
            },
          ],
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      placeholder="Price per Share"
      rejectResponderTermination={true}
      selectionColor="#222"
      style={
        Array [
          Object {
            "marginBottom": 15,
          },
          Object {
            "borderRadius": 7,
          },
          Object {
            "borderColor": "transparent",
            "borderWidth": 0,
          },
          Array [
            Object {
              "fontSize": 40,
            },
            Object {
              "backgroundColor": "#fff",
              "color": "#aaa",
              "height": 45,
              "marginHorizontal": 40,
              "textAlign": "center",
            },
          ],
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#ffa500",
        "borderRadius": 7,
        "height": 45,
        "justifyContent": "center",
        "marginHorizontal": 40,
        "opacity": 0.3,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontWeight": "bold",
            "textAlign": "center",
          },
          Object {
            "fontSize": 40,
          },
        ]
      }
    >
      Calculate
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 40,
          },
          Object {
            "letterSpacing": 1,
            "marginBottom": 30,
            "textTransform": "uppercase",
          },
          Object {
            "color": "#fff",
            "marginHorizontal": 40,
          },
        ]
      }
    >
      Profit: 
      0
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#ff0000",
        "borderRadius": 7,
        "height": 45,
        "justifyContent": "center",
        "marginBottom": 30,
        "marginHorizontal": 40,
        "opacity": 0.3,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontWeight": "bold",
            "textAlign": "center",
          },
          Object {
            "fontSize": 40,
          },
        ]
      }
    >
      Reset All
    </Text>
  </View>
</View>
`;
